src/status.rs

```diff
@@ -103,13 +104,128 @@ fn estimated_total_task_time(
 	}
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub enum ProgressFormat {
+	None,
+	Text,
+	ASCIIBar,
+	HighResBar,
+}
+
+#[derive(Debug)]
+pub struct ParseProgressFormatError {
+	value: String,
+}
+
+impl core::str::FromStr for ProgressFormat {
+	type Err = ParseProgressFormatError;
+	fn from_str(s: &str) -> Result<Self, ParseProgressFormatError> {
+		let s = s.to_lowercase();
+		if s == "none" {
+			Ok(ProgressFormat::None)
+		} else if s == "text" {
+			Ok(ProgressFormat::Text)
+		} else if s == "ascii" {
+			Ok(ProgressFormat::ASCIIBar)
+		} else if s == "highres" {
+			Ok(ProgressFormat::HighResBar)
+		} else {
+			Err(ParseProgressFormatError { value: s })
+		}
+	}
+}
+
+impl fmt::Display for ParseProgressFormatError {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		write!(f, "{}", self.value)
+	}
+}
+
+fn render_progress_bar(width: usize, percentage: f32, text: &String, highres: bool) -> String {
+	while text.len() > width {
+		return String::from(text.split_at(width).0);
+	}
+
+	let textwidth = text.len();
+	let spaceleft = (width - textwidth) / 2;
+	let spaceright = width - textwidth - spaceleft;
+
+	let arrow_width = percentage * width as f32 / 100.;
+	let arrow_full_char = if highres { "▉" } else { "=" };
+	let arrow_cut_char = if highres { "▋" } else { arrow_full_char };
+
+	let arrow_rest = arrow_width - arrow_width.floor();
+	let arrow_rest_char = if arrow_rest < 0.01 {
+		""
+	} else if !highres {
+		">"
+	} else if arrow_rest > 0.84 {
+		"▊"
+	} else if arrow_rest > 0.68 {
+		"▋"
+	} else if arrow_rest > 0.52 {
+		"▌"
+	} else if arrow_rest > 0.36 {
+		"▍"
+	} else if arrow_rest > 0.20 {
+		"▎"
+	} else {
+		"▏"
+	};
+
+	// If highres, the arrow characters are all 3 bytes. This is assumed below.
+	let arrow_charlen = if highres { 3 } else { 1 };
+	assert!(arrow_full_char.len() == arrow_charlen);
+	assert!(arrow_rest_char.is_empty() || arrow_rest_char.len() == arrow_charlen);
+
+	let arrow = arrow_full_char.repeat(arrow_width.floor() as usize) + arrow_rest_char;
+	let arrowlen = arrow.chars().count();
+
+	let left = if arrowlen < spaceleft {
+		format!("{}{}", arrow, " ".repeat(spaceleft - arrowlen))
+	} else {
+		format!(
+			"{}{}",
+			arrow.split_at((spaceleft - 1) * arrow_charlen).0,
+			arrow_cut_char
+		)
+	};
+	let lefttextlen = spaceleft + textwidth;
+	let right = if arrowlen < lefttextlen {
+		" ".repeat(spaceright)
+	} else if arrowlen == width {
+		format!(
+			"{}{}",
+			arrow
+				.split_at(lefttextlen * arrow_charlen)
+				.1
+				.split_at((spaceright - 1) * arrow_charlen)
+				.0,
+			arrow_cut_char
+		)
+	} else {
+		format!(
+			"{}{}",
+			arrow.split_at(lefttextlen * arrow_charlen).1,
+			" ".repeat(width - arrowlen)
+		)
+	};
+
+	assert!(left.chars().count() == spaceleft);
+	assert!(right.chars().count() == spaceright);
+	assert!(spaceleft + text.chars().count() + spaceright == width);
```

Comment:
There's also `debug_assert`. `assert` still checks in release mode (unlike in C and C++).